File: game-server/src/main/java/org/coinen/reactive/pacman/controller/socket/io/PlayerController.java
package org.coinen.reactive.pacman.controller.socket.io;

import com.google.protobuf.Empty;
import org.coinen.pacman.Location;
import org.coinen.pacman.Player;
import org.coinen.reactive.pacman.service.PlayerService;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

public class PlayerController {

    final PlayerService playerService;

    public PlayerController(PlayerService playerService) {
        this.playerService = playerService;
    }

    public Mono<Empty> locate(Flux<Location> messages) {
        return playerService.locate(messages)
            .thenReturn(Empty.getDefaultInstance());
    }

    public Flux<Player> players() {
        return playerService.players()
                            .onBackpressureDrop();
    }
}

File: game-server/src/main/typescript/controller/rsocket/PlayerController.ts
import * as rsocket_flowable from 'rsocket-flowable';
import * as google_protobuf_empty_pb from 'google-protobuf/google/protobuf/empty_pb';
import {Location, Player, RSocketRPCServices} from "game-idl";
import {PlayerService} from '../../service';
import {Disposable} from 'reactor-core-js';
import FlowableAdapter from './support/FlowableAdapter';
import {Flux} from 'reactor-core-js/flux';

export class PlayerController implements RSocketRPCServices.PlayerService {
    constructor(private playerService: PlayerService, private uuid: string) {}

    locate(message: rsocket_flowable.Flowable<Location>, metadata?: Buffer): rsocket_flowable.Single<google_protobuf_empty_pb.Empty> {
        return new rsocket_flowable.Flowable(subject => {
            let disposable: Disposable = {
                dispose: () => {}
            };

            subject.onSubscribe({
                request: () => {},
                cancel: () => disposable.dispose()
            });

            disposable = this.playerService.locate(this.uuid, Flux.from(FlowableAdapter.wrap(message as any)))
                .consume(
                    () => {},
                    (e: Error) => subject.onError(e),
                    () => subject.onComplete()
                )
        }) as any;
    }

    players(message: google_protobuf_empty_pb.Empty, metadata?: Buffer): rsocket_flowable.Flowable<Player> {
        return this.playerService.players() as any;
    }

}
